src/closure.cc

```diff
@@ -97,6 +112,21 @@ void Closure::Marshal(GClosure     *base,
     #endif
 }
 
+void Closure::Marshal(GClosure     *base,
+                      GValue       *g_return_value,
+                      uint          n_param_values,
+                      const GValue *param_values,
+                      gpointer      invocation_hint,
+                      gpointer      marshal_data) {
+
+    auto closure = (Closure *) base;
+    auto signal_id = GPOINTER_TO_UINT(marshal_data);
+
+    AsyncCallEnvironment* env = reinterpret_cast<AsyncCallEnvironment *>(AsyncCallEnvironment::asyncHandle.data);
+    env->Call([&]() {
+        Closure::Execute(closure->info, signal_id, closure->persistent, g_return_value, n_param_values, param_values);
+    });
+}
```

Comment:
Would you mind explaining why it's now always wrapped in an `AsyncCallEnvironment::Call` instead of directly calling when it's in the same thread?

Reply:
I thought it would good to avoid code duplication. But it is really not necessary to wrap the call into a function when we are on the same thread. So I rewrote it a bit.